moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp

```diff
@@ -0,0 +1,1012 @@
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <limits>
+//#include <fstream>
+//#include <iostream>
+#include <Eigen/Geometry>
+#include <algorithm>
+#include <angles/angles.h>
+#include <cmath>
+#include <console_bridge/console.h>
+#include <moveit/trajectory_processing/time_optimal_trajectory_generation.h>
+#include <vector>
+
+namespace trajectory_processing
+{
+class LinearPathSegment : public PathSegment
+{
+public:
+  LinearPathSegment(const Eigen::VectorXd& start, const Eigen::VectorXd& end)
+    : start(start), end(end), PathSegment((end - start).norm())
+  {
+  }
+
+  Eigen::VectorXd getConfig(double s) const
+  {
+    s /= length;
+    s = std::max(0.0, std::min(1.0, s));
+    return (1.0 - s) * start + s * end;
+  }
+
+  Eigen::VectorXd getTangent(double /* s */) const
+  {
+    return (end - start) / length;
+  }
+
+  Eigen::VectorXd getCurvature(double /* s */) const
+  {
+    return Eigen::VectorXd::Zero(start.size());
+  }
+
+  std::list<double> getSwitchingPoints() const
+  {
+    return std::list<double>();
+  }
+
+  LinearPathSegment* clone() const
+  {
+    return new LinearPathSegment(*this);
+  }
+
+private:
+  Eigen::VectorXd start;
+  Eigen::VectorXd end;
+};
+
+class CircularPathSegment : public PathSegment
+{
+public:
+  CircularPathSegment(const Eigen::VectorXd& start, const Eigen::VectorXd& intersection, const Eigen::VectorXd& end,
+                      double maxDeviation)
+  {
+    if ((intersection - start).norm() < 0.000001 || (end - intersection).norm() < 0.000001)
+    {
+      length = 0.0;
+      radius = 1.0;
+      center = intersection;
+      x = Eigen::VectorXd::Zero(start.size());
+      y = Eigen::VectorXd::Zero(start.size());
+      return;
+    }
+
+    const Eigen::VectorXd startDirection = (intersection - start).normalized();
+    const Eigen::VectorXd endDirection = (end - intersection).normalized();
+
+    if ((startDirection - endDirection).norm() < 0.000001)
+    {
+      length = 0.0;
+      radius = 1.0;
+      center = intersection;
+      x = Eigen::VectorXd::Zero(start.size());
+      y = Eigen::VectorXd::Zero(start.size());
+      return;
+    }
+
+    const double startDistance = (start - intersection).norm();
+    const double endDistance = (end - intersection).norm();
+
+    double distance = std::min((start - intersection).norm(), (end - intersection).norm());
+    const double angle = acos(startDirection.dot(endDirection));
+
+    distance = std::min(distance, maxDeviation * sin(0.5 * angle) / (1.0 - cos(0.5 * angle)));  // enforce max deviation
```

Comment:
if angle = 0 this will be a divide by zero error - add check for this and maybe set distance = 0 if angle = 0?